BufferPool.java

```diff
@@ -0,0 +1,245 @@
+    if (0 == bufferPoolSize) {
+      throw new IOException(
+          String.format("The total size of the buffer [%d] is " +
+                  "smaller than a single block [%d]."
+                  + "please consider increase the buffer size " +
+                  "or decrease the block size",
+              memoryBufferLimit, this.singleBufferSize));
+    }
+    this.bufferPool = new LinkedBlockingQueue<>(bufferPoolSize);
+    for (int i = 0; i < bufferPoolSize; i++) {
+      this.bufferPool.add(ByteBuffer.allocateDirect(
+          (int) this.singleBufferSize));
+    }
+
+    this.isInitialize.set(true);
+  }
+
+  /**
+   * Check if the buffer pool has been initialized.
+   *
+   * @throws IOException if the buffer pool is not initialized
+   */
+  private void checkInitialize() throws IOException {
+    if (!this.isInitialize.get()) {
+      throw new IOException(
+          "The buffer pool has not been initialized yet");
+    }
+  }
+
+  /**
+   * Obtain a buffer from this buffer pool through the method.
+   *
+   * @param bufferSize expected buffer size to get
+   * @return a buffer wrapper that satisfies the bufferSize.
+   * @throws IOException if the buffer pool not initialized,
+   *                     or the bufferSize parameter is not within
+   *                     the range[1MB to the single buffer size]
+   */
+  public ByteBufferWrapper getBuffer(int bufferSize) throws IOException {
+    this.checkInitialize();
+    if (bufferSize > 0 && bufferSize <= this.singleBufferSize) {
+      ByteBufferWrapper byteBufferWrapper = this.getByteBuffer();
+      if (null == byteBufferWrapper) {
+        // Use a disk buffer when the memory buffer is not enough
+        byteBufferWrapper = this.getMappedBuffer();
+      }
+      return byteBufferWrapper;
+    } else {
+      String exceptionMsg = String.format(
+          "Parameter buffer size out of range: 1048576 to %d",
+          this.singleBufferSize
+      );
+      throw new IOException(exceptionMsg);
+    }
+  }
+
+  /**
+   * Get a ByteBufferWrapper from the buffer pool.
+   *
+   * @return a new byte buffer wrapper
+   * @throws IOException if the buffer pool is not initialized
+   */
+  private ByteBufferWrapper getByteBuffer() throws IOException {
+    this.checkInitialize();
+    ByteBuffer buffer = this.bufferPool.poll();
+    return buffer == null ? null : new ByteBufferWrapper(buffer);
+  }
+
+  /**
+   * Get a mapped buffer from the buffer pool.
+   *
+   * @return a new mapped buffer
+   * @throws IOException If the buffer pool is not initialized.
+   *                     or some I/O error occurs
+   */
+  private ByteBufferWrapper getMappedBuffer() throws IOException {
+    this.checkInitialize();
+    File tmpFile = File.createTempFile(Constants.BLOCK_TMP_FILE_PREFIX,
+        Constants.BLOCK_TMP_FILE_SUFFIX, this.diskBufferDir);
+    tmpFile.deleteOnExit();
+    RandomAccessFile raf = new RandomAccessFile(tmpFile, "rw");
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *raf*. Currently, there are execution paths that do not contain closure statements, e.g., when exception is thrown by FileChannel.map. Either a) close *raf* in a try-finally block or b) close the resource by declaring *raf* in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).